File: src/switch_tracker/core/concurrency.py
"""Bounded, order-preserving concurrency."""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable, Sequence


async def bounded_gather[T, R](
    items: Sequence[T],
    limit: int,
    worker: Callable[[T], Awaitable[R]],
) -> list[R]:
    """Run ``worker`` over ``items``, at most ``limit`` in flight.

    Input order is preserved in the result regardless of completion order --
    probe prints a table, and rows arriving in completion order would shuffle
    it between runs for no reason.

    Stores are independent hosts, so running them concurrently makes a run
    cost roughly the slowest store rather than the sum of all of them. The
    per-host throttle in http.py already serialises requests WITHIN a host,
    so this bound is about machine resources, not politeness.
    """
    if not items:
        return []

    semaphore = asyncio.Semaphore(limit)

    async def guarded(item: T) -> R:
        async with semaphore:
            return await worker(item)

    return list(await asyncio.gather(*(guarded(item) for item in items)))

File: src/core/concurrency.ts
/**
 * Runs `worker` over `items` with at most `limit` in flight at once,
 * preserving input order in the returned array regardless of completion
 * order. Used to process independent stores (different hosts) concurrently
 * — the politeness throttle in http.ts already serializes requests within
 * the same host, so serializing *across* stores on top of that only made
 * every run take the sum of every store's latency instead of the slowest
 * one's.
 */
export async function mapWithConcurrency<T, R>(
  items: readonly T[],
  limit: number,
  worker: (item: T, index: number) => Promise<R>,
): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let next = 0;

  async function runNext(): Promise<void> {
    const i = next++;
    if (i >= items.length) return;
    results[i] = await worker(items[i] as T, i);
    return runNext();
  }

  await Promise.all(Array.from({ length: Math.min(limit, items.length) }, runNext));
  return results;
}
